mixer/pkg/runtime2/routing/string.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+
+	tpb "istio.io/api/mixer/v1/template"
+)
+
+func (t *Table) String() string {
+	var b bytes.Buffer
+
+	fmt.Fprintln(&b, "[Routing ExpectedTable]")
```

Comment:
if we don't need `fmt`, why not just `b.WriteString("[Routing ExpectedTable]")` (and perhaps similar below for other bits)?

Reply:
i just recall early flame graphs of mixer execution being heavily dominated by `fmt` calls. slightly worried about impact on perf. was wondering if we could avoid unnecessary printf's etc in places. just thinking out loud.

Reply:
I believe all fmt calls are in String() method. This is not called in the main call path.